### PR TITLE
🎉 (dod) support dods in axis labels

### LINKED ISSUE #2484: Support details on demand in the X and Y axes' display names
image

- [ ] Support DoDs in both axes' display names
- [ ] Validate DoDs so that a chart can't be saved if its axes' display names reference non-existent IDs
- [ ] Ensure these DoDs get appended to the footnotes of a static export correctly (with superscript reference numbers too)

### LINKED ISSUE #3009: Footer text overlaps in some Grapher thumbnails
## Description

Examples:
- ~https://ourworldindata.org/grapher/thumbnail/global-warming-by-gas-and-source~ ok now
- ~https://ourworldindata.org/grapher/thumbnail/total-affected-by-natural-disasters~ ok now
- ~https://ourworldindata.org/grapher/thumbnail/homicide-rate-unodc~ ok now
- https://ourworldindata.org/grapher/thumbnail/global-warming-by-gas-and-source?nocache&imWidth=525&imHeight=400

## Screenshots

Screenshot 2024-02-23 at 14 02 52

### PR DESCRIPTION
**Pull request by sophiamersmann**
**tldr** Adds support for details on demand in x- and y-axis labels.

This PR allows authors to use DoD syntax in the admin's x- and y-label fields. Some chart types (Scatter, Marimekko) use the indicator's display name as default if no custom axis label is given. We toyed with the idea of automatically checking for details with a specific prefix (e.g. `grapher_indicator_per-capita-emissions`) when the metadata default is used but decided against it for now. The automatic approach for default labels is implemented though but temporarily disabled (see the [last commit](https://github.com/owid/owid-grapher/pull/3150/commits/e9da915dc47d78984433b24fe46497440fccafbe)), so that we can quickly enable it when the time is right.

### Summary

- This PR adds support for details on demand in axis labels by:
    - allowing authors to use Dod syntax in the admin's x- and y-axis label fields
    - (disabled feature) automatically checking for an existing detail with a specific prefix when a default label is used
- Details are validated in the admin
    - If invalid details are specified in the x- and y-axis label fields, an error message is displayed, and the chart cannot be saved
    - If details are specified in the dimension's display name field, an error message is displayed, and the chart cannot be saved

### Example

On staging: http://staging-site-dod-in-axis-labels/admin/charts/937/edit (the deploy queue seems to stuck for some reason, so you won't see DoDs on the grapher page)

### Details

- How does Grapher know which label is currently in use?
    - Scatter plots and Marimekko charts define `defaultYAxisLabel` and `defaultXAxisLabel`
    - If no custom overwrite is provided and `default{X,Y}AxisLabel` is given, Grapher knows that a default label is in use
- How does Grapher connect a given indicator name to an existing detail?
    - If the indicator name is used as default label, Grapher checks if a detail with id `grapher_indicator_my-slugified-display-name` exists (e.g. `grapher_indicator_biofuels`)
    - If it does exist, it replaces the indicator name with Dod syntax just before rendering, i.e. the display name "Biofuels" is turned into `[Biofuels](#dod:grapher_indicator_biofuels)`
- How does Grapher decide how to highlight a given Dod?
    - The logic used to be:
        - If rendered as HTML, we're in an interactive environment, so underline the detail
        - If rendered as SVG, we're exporting to a static chart, so show a reference number (or don't show anything if details shouldn't be included in the export)
    - That doesn't work for text that is always rendered to SVG (like axis labels); we need to be more explicit here (see Grapher's `detailsMarkerInSvg` property)

### Notes

- Details used to be fetched late (in the footer). That was okay since we only needed details after an interaction had taken place (e.g. the user hovers over a detail or clicks the download button). With these changes, details need to be validated early, which means that the details object must exist when Grapher loads. To guarantee that it does, details are now loaded in the `<Head />` component
- SVG text doesn't support dotted underlines, so MarkdownTextWrap manually draws dotted lines that act as underlines for details
- Since I updated the interface of `renderSVG` of the MarkdownTextWrap class to account for `dodMarker`, I did the same for the TextWrap class (it makes the interface a bit more involved, but I value consistency higher)

### Caveats

- Technically, axis labels now support any Markdown syntax, although we wouldn't want authors to use bold or italic text in axis labels for consistency. We could solve this on a technical level by adding a new `DetailsTextWrap` class or maybe by allowing/blocking specific Markdown syntax. I haven't done that though – I hope that a convention of not using Markdown syntax for axis labels is enough. But I went back and forth on this – what do you think, Ike?

### Related issues

- Closes #2484.
- Also fixes #3009 (to place the underline correctly below the DoD, I fixed some bugs in `MarkdownTextWrap`; apparently, that has also fixed alignment issues for thumbnails 🙏🏻 )

The SVG tester flags many charts since axis labels are now wrapped in a group element.

The code scanner complains about running `slugify` on invalidated input (since it uses a regex), I guess because we're running it on an indicator's display name. But indicator metadata either comes from the ETL or from our internal admin, so it should be fine? Don't know how to let the code scanner know, though.

### CODE REVIEW

**Comment by sophiamersmann** [comment]
Current dependencies on/for this PR:
* `master`
  * **PR #3150** <a href="https://app.graphite.dev/github/pr/owid/owid-grapher/3150?utm_source=stack-comment-icon" target="_blank">Graphite</a>  👈

This [stack of pull requests](https://stacking.dev/?utm_source=stack-comment) is managed by [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3150?utm_source=stack-comment).

**Comment by sophiamersmann** [comment]
I'm seeing [this](https://github.com/owid/owid-grapher/issues/2745) now, and am wondering if – instead of allowing dod syntax for an indicator's display name – we should rather look for dods automatically based on some naming scheme:

> To avoid polluting our metadata with DoD syntax, it might be best to try to automatically add DoDs if we see there's a definition for the term with a specific prefix e.g. grapher_category_conflict_and_terrorism or grapher_unit_lumen-hours

**Comment by sophiamersmann** [comment]
Hey @ikesau, following my previous comment 👆🏻, I have re-written parts of the PR to automatically check for DoDs based on a naming scheme (the PR description is updated accordingly). I have just discussed this with Christian and Daniel in the data viz meeting. They're on board, but Daniel is going to check in with Lars and Joe about this. I'm removing my request for review for now because I don't want you to review something that might be revised.

File: site/detailsOnDemand.tsx [hunk 1]
```diff
@@ -17,18 +17,20 @@ declare global {
...
-        `${BAKED_BASE_URL}/dods.json`,
-        {
-            method: "GET",
-            credentials: "same-origin",
-            headers: {
-                Accept: "application/json",
-            },
-        }
-    ).then((res) => res.json())
+    if (!window.details) {
```
Comment (ikesau) [inline site/detailsOnDemand.tsx:20]: I guess there's a chance we make this request twice now if the first request doesn't resolve by the time this fires. Not sure how likely that'll be, but I wonder if there's something we can do about it. 🤔

**Review by ikesau — APPROVED** [review]
Nice! Just style nit picks from me, ignore them as you please 😎 

I tested each chart type with details and duplicates in all places, and it all seemed to work fine.

Didn't test the prefix codepath, let me know if you think I should 👍

File: adminSiteClient/ChartEditorPage.tsx [hunk 2]
```diff
@@ -278,12 +289,71 @@ export class ChartEditorPage
...
+                    ] = `Invalid detail(s) specified: ${references.join(", ")}`
+                }
+            }
+        )
+
+        return errorMessages
+    }
+
+    @computed
+    get errorMessagesForDimensions(): ChartEditorManager["errorMessagesForDimensions"] {
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply adminSiteClient/ChartEditorPage.tsx:330]: Yes, indicator names are not only used as default axis labels, they're used across our UI. If we allowed DoD syntax here, the raw DoD syntax string would show up in all kinds of places (e.g. the table). That's true for most of our text properties that don't support DoDs but what makes this special is that adding detail syntax to an indicator name would appear to work for axis labels, i.e. an axis label would actually render a DoD, but it would be broken in the sense that we don't keep track of DoDs in indicator names, thus the DoD wouldn't be appended to static charts, for example.

File: packages/@ourworldindata/grapher/src/header/HeaderManager.ts [hunk 3]
```diff
@@ -8,7 +9,7 @@ export interface HeaderManager {
     logo?: string
     canonicalUrl?: string
     tabBounds?: Bounds
-    detailsOrderedByReference?: Set<string>
```
[the remark this answers is not in the document]
Comment (sophiamersmann) [reply packages/@ourworldindata/grapher/src/header/HeaderManager.ts:11]: It felt a little weird to me to use a set, which is usually thought of as an unordered collection, for a data structure where the order is important. I guess in JS it doesn't make much of a difference, but we also usually use `detailsOrderedByReference` as an array, so it felt more appropriate to me.

File: site/detailsOnDemand.tsx (hunk 1 above)
Comment (sophiamersmann) [reply site/detailsOnDemand.tsx:20]: If the `Head` component is used on all our pages, we could just remove the second request? We could also await the first request but to make that request blocking is probably too much of a compromise. Thinking about it, if the first request does take forever, we'll probably notice because Grapher's DoDs won't work correctly...

**Comment by sophiamersmann** [comment]
Thanks for the detailed comments 👌🏻

**Comment by sophiamersmann** [comment]
### Merge activity

* **Mar 7, 3:46 AM EST**: @sophiamersmann started a stack merge that includes this pull request via [Graphite](https://app.graphite.dev/github/pr/owid/owid-grapher/3150).
